### PR TITLE
feat(lua): vim.tbl_contains supports general tables and predicates

### LINKED ISSUE #23024: Extend `vim.tbl_contains` to nested tables
### Problem

As `vim.tbl_contains(table, value)` uses `table[i] == value`, it cannot be used to check for values that are itself tables. (Lua's `==` checks for identity, not equality.) Example:
```lua
foo = { { 'bar', 'baz' }}
print(vim.tbl_contains(foo, { 'bar', 'baz' })
-- false
```

This is needed, e.g., in https://github.com/neovim/neovim/pull/22784

### Expected behavior

Code snippet should return `true`, either by

1. adding a new function `vim.tbl_deep_contains` that uses `deep_equal` instead of `==`;
2. adding an `opts` table to  `vim.tbl_contains` with boolean key `deep_equal` (default `false`);
3. adding a new function `vim.tbl_contains_func` where `value` can be user-supplied predicate.
(Note that `value` can be anything, even a funcref, so we can't dispatch on type)
4. adding an `opts` table to `vim.tbl_contains` with funcref `predicate` key (in which case `value` is ignored or validated to be `nil`) -- or boolean `funcref` key in which case `value` is validated to be a funcref?

### PR DESCRIPTION
**Pull request by clason**
Problem: `vim.tbl_contains` only works for list-like tables (integer keys without gaps) and primitive values (in particular, not for nested tables).

Solution: Rename `vim.tbl_contains` to `vim.list_contains` and add new `vim.tbl_contains` that works for general tables and optionally allows `value` to be a predicate function that is checked for every key.

Closes #23024

### CODE REVIEW

File: runtime/doc/lua.txt [hunk 1]
```diff
@@ -1797,16 +1810,31 @@ tbl_add_reverse_lookup({o})                     *vim.tbl_add_reverse_lookup()*
     Return: ~
         (table) o
 
-tbl_contains({t}, {value})                                *vim.tbl_contains()*
-    Checks if a list-like (vector) table contains `value`.
+tbl_contains({t}, {value}, {opts})                        *vim.tbl_contains()*
+    Checks if a table contains a given key, specified either directly or via a
+    predicate that is checked for each key.
```
Comment (zeertzjq) [inline runtime/doc/lua.txt:1815]:
```suggestion
    Checks if a table contains a given value, specified either directly or via a
    predicate that is checked for each value.
```

Comment (zeertzjq) [inline runtime/doc/lua.txt:1815]:
```suggestion
    Checks if a table contains a given value, specified either directly or via a
    predicate that is checked for each value.
```

File: runtime/doc/lua.txt [hunk 2]
```diff
@@ -1698,6 +1698,19 @@ is_callable({f})                                           *vim.is_callable()*
...
+
+    Parameters: ~
+      • {t}      (table) Table to check (must be list-like, not validated)
+      • {value}  any Value to compare
+
+    Return: ~
+        (boolean) `true` if `t` contains `value`
+
+    See also: ~
+      • |vim.tbl_contains()| for checking keys in general tables
```
Comment (zeertzjq) [inline runtime/doc/lua.txt:1712]:
```suggestion
      • |vim.tbl_contains()| for checking values in general tables
```

**Comment by mfussenegger** [comment]
> I have not changed any of the current uses of `tbl_contains` back to the (faster) `list_contains`. Should I do so? Are there any that look like they're not "list-safe"? @mfussenegger (for codelens)

The check in codelens is list-safe. 

E.g.:

```
lua =vim.lsp.get_active_clients()[1].server_capabilities.executeCommandProvider
```

With lua-ls returns:

```
{
  commands = { "lua.removeSpace", "lua.solve", "lua.jsonToLua", "lua.setConfig", "lua.getConfig", "lua.autoRequire" }
}
```

**Review by mfussenegger — COMMENTED** [review]
Just to have mentioned it: Instead of a `list_contains` it could also be an option to bring this into the iterators of https://github.com/neovim/neovim/pull/23029

**Comment by clason** [comment]
Merging this now to unblock #22784. Once `vim.iter` is merged, we need to revisit the API anyway to see what we want to keep as a convenience function on top of it and what can be replaced by an explicit iterator.

**Comment by justinmk** [comment]
> Just to have mentioned it: Instead of a `list_contains` it could also be an option to bring this into the iterators of #23029

Yes please :)

**Pull request merged**